### DISCUSSION TITLE
Please introduce nonexcept concept in C#

### DISCUSSION THREAD

**Discussion by IrakliLomidze** [dotnet/csharplang#5807] [Language Ideas]
Please introduce C++ nonexcept concept in C#.

some times code written is safe way and we may be tolerance to non throwing exceptions.
AFAIK  exceptions are expensive. It would be grate to have an options to have block of the performance code that not throwing exceptions.

idea to have keyword on nonexcept on methods.

**Comment by jnm2** [comment]
How would the compiler prevent exceptions from being thrown by a method with a `nonexcept` modifier? If the method calls another method, including an interface method or delegate, it would also have to be annotated as nonexcept or else it's possible that the method would still throw.

Also, how are exceptions like division by zero or NREs handled? What does the compiler do here?

```cs
public nonexcept void M(int n, int? o, string? p)
{
    var x = 100 / n; // Might throw DivideByZeroException
    var y = n + 1; // Might throw OverflowException
    var z = o.Value; // Might throw InvalidOperationException
    var w = p.Length; // Might throw NullReferenceException
}
```

Also, how does the compiler know that LINQ methods can ever be used in a nonexcept method? Even LINQ methods that don't throw for non-null inputs, like .Select and .Where, themselves are generically invoking `IEnumerator<T>` methods and those interface methods can't be annotated as `nonexcept`, so `.Select` and `.Where` can't be annotated as `nonexcept`, so they can't be used inside a `nonexcept` method. And it just gets worse in every direction from here.

One way out might be for the compiler to give up and force you to wrap every `nonexcept` method's body in `try`...`catch (Exception ex)` which is not only an antipattern but also does nothing to help with the motivation of avoiding exception expense. This would maximize the expense.

**Reply by HaloFour** [reply]
I'm not terribly familiar with the `noexcept` modifier in C++ but from a quick read of the [spec](https://en.cppreference.com/w/cpp/language/noexcept_spec):

> Non-throwing functions are permitted to call potentially-throwing functions. Whenever an exception is thrown and the search for a handler encounters the outermost block of a non-throwing function, the function [std::terminate](https://en.cppreference.com/w/cpp/error/terminate) or [std::unexpected](https://en.cppreference.com/w/cpp/error/unexpected) (until C++17) is called.

So the method itself is guaranteed to not throw by only by virtue that any exception thrown is immediately caught and effectively fails fast.  So a C# equivalent might be:

```cs
public nonexcept void M(int n, int? o, string? p) {
    var x = 100 / n; // Might throw DivideByZeroException
    var y = n + 1; // Might throw OverflowException
    var z = o.Value; // Might throw InvalidOperationException
    var w = p.Length; // Might throw NullReferenceException
}

// translated into:
public void M(int n, int? o, string? p) {
    try {
        var x = 100 / n; // Might throw DivideByZeroException
        var y = n + 1; // Might throw OverflowException
        var z = o.Value; // Might throw InvalidOperationException
        var w = p.Length; // Might throw NullReferenceException
    }
    catch (Exception exception) {
         Environment.FailFast("Something something nonexcept", exception);
    }
}

```

Shame source generators can't do AOP, seems like it would be easy to accomplish with:

```cs
[NoExcept]
public partial void M(...) { ... }
```

**Reply by jnm2** [reply]
Thanks, that helps. It seems to indicate that there's not a way to achieve OP's goal:

> AFAIK exceptions are expensive. It would be grate to have an options to have block of the performance code that not throwing exceptions.

**Reply by qrli** [reply]
I don't think `noexcept` fits into .NET either.
There are a few different use cases of `noexcept`, e.g.:
* When writing a callback from C or COM code which does not understand C++ exceptions, it helps to make sure the function won't throw. Compiler could produce warning to help. 
  * However, in .NET, PInvoke already handles the marshaling of exceptions.
* For realtime or embedded scenarios, you may want to avoid the cost of exception handling and enforce C-style error handling.
  * If that's really critical, .NET may be the wrong choice because you cannot avoid GC pauses in the 1st place.
* Hot code path optimization
  * It only helps at that compiler might help find some unintended possible exceptions. `noexcept` is more of a declaration of intention. Programmer need to ensure it does not raise at runtime (at extreme, `noexcept` does not prevent OutOfMemory or StackOverflow, it is for devs to make sure they won't happen normally), which would be the same with/out it.

**Comment by IrakliLomidze** [comment]
if there are members from MSFT :
is nonexcept a matter of runtime engine or compiler?

**Reply by bernd5** [reply]
The compiler (roslyn) emits IL which is translated to platform specific native code at runtime.
Therefore the compiler can't do a lot here.
In addition I don't think introducing a new keyword is it worth - but I could imagine that the runtime could introduce a new Attribute (e.g. `[NeverThrows]` or `[NoExcept]`) which might be used to annotate those methods. The runtime/JIT could then make use of the annotation and optimize the codegen (if possible).

Take for example: https://sharplab.io/#v2:EYLgtghglgdgNAFxBAzmAPgAQEwAYCwAUDgIxFGwIAEUVAvFQIIAOzAdAMIQA2AxgBQBKANzli2JqwDeRKnKqYAzApIA2GjGpc+QmYXkGFAdioARAPbaBI2fIC+tuY4XLMajdQtXdzwwgAWAE7mAO5UMACu3NzC8s4OhHZAA

I don't see how the compiler could do any better with the annotation. 

In addition an analyzer can make checks if the method really never throws.

**Reply by Joe4evr** [reply]
> In addition an analyzer can make checks if the method really never throws.

Even that wouldn't accomplish much because the runtime is required by ECMA-335 to throw certain exceptions for certain situations regardless of the method body (`I.12.4.2.1`).
> * System.ArithmeticException
> * System.DivideByZeroException
> * System.ExecutionEngineException
> * System.InvalidAddressException
> * System.OverflowException
> * System.SecurityException
> * System.StackOverflowException
> * System.TypeLoadException
> * System.IndexOutOfRangeException
> * System.InvalidAddressException
> * System.InvalidCastException
> * System.MissingFieldException
> * System.MissingMethodException
> * System.NullReferenceException
> * System.OutOfMemoryException

No amount of static analysis will prevent all of these.

> AFAIK exceptions are expensive

Only if one is actually *thrown*. If you want to avoid the cost of having an exception thrown, the best thing is to do your up-front checks so that the inner workings of your code don't have to.

**Comment by IrakliLomidze** [comment]
"Only if one is actually thrown. If you want to avoid the cost of having an exception thrown, the best thing is to do your up-front checks so that the inner workings of your code don't have to"


Are you absolutely sure, that code execution in try block and without it are the same ?
{ some code }
try {some code } catch {}

**Comment by IrakliLomidze** [comment]
Please see the result, how 
{ some code }
try {some code } catch {}
is different

https://sharplab.io/#v2:EYLgtghglgdgNAFxBAzmAPgAQEwAYCwAUDgIxHmFE4AEAggA70DeR1b1mAzByQGzWwE1AMIQANgGMSACkECAlC0LsV1BACcAnq1Vslu3VGoBeAdQBU1bAG4dugL53VEiAgkALatICiADwkApvQIUAD2MNQBvvJOKvoGKgju6qEA7pG+1uyx7I7KBpgA7AK2+Wx5Kk5cPPxyopLYsjBCUDFl1PGGJmaWNtQ5bEUlTnn2QA===
